### ISSUE TITLE
Azure Semaphore implementation

### ISSUE THREAD

**Issue by andysleigh** [madelson/DistributedLock#205]
I was looking into using this library for it's semaphore functionality, and I noticed there's currently no Azure implementation of IDistributedSemaphore. Does this exist, and I'm not just looking in the right place? If not is this something you would accept a PR for if I was able to implement it?

**Comment by madelson** [comment]
@andysleigh you are correct; there is currently not Azure semaphore implementation. I would definitely love to have one contributed so long as a robust implementation can reasonably be achieved.

If you'd like to contribute this, a starting point would be coming up with a basic design sketch and posting it here so we can make sure we're in the same page before you sink more effort into the implementation.
